File: sendjob_ipython.py
# -*- coding: utf-8 -*-

from optparse import OptionParser
import os
import DrQueue
from DrQueue import Job as DrQueueJob
from DrQueue import Client as DrQueueClient


def main():
    # parse arguments
    parser = OptionParser()
    parser.usage = "%prog [options] -n name -r renderer -f scenefile"
    parser.add_option("-s", "--startframe",
                      dest="startframe", default=1, help="first frame")
    parser.add_option("-e", "--endframe",
                      dest="endframe", default=1, help="last frame")
    parser.add_option("-b", "--blocksize",
                      dest="blocksize", default=1, help="size of block")
    parser.add_option("-n", "--name",
                      dest="name", default=None, help="name of job")
    parser.add_option("-r", "--renderer",
                      dest="renderer", help="render type (maya|blender|mentalray)")
    parser.add_option("-f", "--scenefile",
                      dest="scenefile", default=None, help="path to scenefile")
    parser.add_option("-o", "--options",
                      dest="options", default="{}", help="specific options for renderer as Python dict")
    parser.add_option("-w", "--wait",
                      action="store_true", dest="wait", default=False, help="wait for job to finish")
    parser.add_option("-v", "--verbose",
                      action="store_true", dest="verbose", default=False, help="verbose output")
    (options, args) = parser.parse_args()

    # initialize DrQueue client
    client = DrQueueClient()

    # initialize DrQueue job
    job = DrQueueJob(options.name, int(options.startframe), int(options.endframe), int(options.blocksize), options.renderer, options.scenefile, eval(options.options))

    # run job with client
    client.job_run(job)

    # tasks which have been created
    tasks = client.query_tasks_of_job(job['name'])

    # wait for all tasks of job to finish
    if options.wait:
        for task in tasks:
            ar = client.task_wait(task['msg_id'])
            # add some verbose output
            if options.verbose:
                cpl = ar.metadata.completed
                msg_id = ar.metadata.msg_id
                status = ar.status
                engine_id = ar.metadata.engine_id
                print("Task %s finished with status '%s' on engine %i at %i-%02i-%02i %02i:%02i:%02i." % (msg_id, status, engine_id, cpl.year, cpl.month, cpl.day, cpl.hour, cpl.minute, cpl.second))
                if ar.pyerr != None:
                    print(ar.pyerr)
        print("Job %s finished." % job['name'])

if __name__ == "__main__":
    main()

File: ruby/sendjob_ipython.rb
require 'optparse'
require "rubypython" 
RubyPython.start(:python_exe => "python2.6")
sys = RubyPython.import "sys" 
sys.argv = [""]
DrQueue = RubyPython.import("DrQueue")


# parse arguments
options = {}
options[:startframe] = 1
options[:endframe] = 1
options[:blocksize] = 1
options[:options] = "{}"
options[:retries] = 1
options[:owner] = Etc.getlogin
options[:wait] = false
options[:verbose] = false

OptionParser.new do |opts|
    opts.banner = "Usage: "+__FILE__+" [options] -n name -r renderer -f scenefile"
    opts.on("-s N", "--startframe N", Integer, "first frame") do |startframe|
        options[:startframe] = startframe
    end
    opts.on("-e N", "--endframe N", Integer, "last frame") do |endframe|
        options[:endframe] = endframe
    end
    opts.on("-b N", "--blocksize N", Integer, "size of block") do |blocksize|
        options[:blocksize] = blocksize
    end
    opts.on("-n NAME", "--name NAME", String, "name of job") do |name|
        options[:name] = name
    end
    opts.on("-r RENDERER", "--renderer RENDERER", String, "render type (maya|blender|mentalray)") do |renderer|
        options[:renderer] = renderer
    end
    opts.on("-f FILE", "--scenefile FILE", String, "path to scenefile") do |scenefile|
        options[:scenefile] = scenefile
    end
    opts.on("-o OPTS", "--options OPTS", String, "specific options for renderer as Python dict") do |extraoptions|
        options[:options] = extraoptions
    end
    opts.on("--retries N", Integer, "number of retries for every task") do |retries|
        options[:retries] = retries
    end
    opts.on("--owner OWNER", String, "Owner of job. Default is current username.") do |owner|
        options[:owner] = owner
    end
    opts.on("-w", "--wait", "wait for job to finish") do |wait|
        options[:wait] = true
    end
    opts.on("-v", "--verbose", "verbose output") do |verbose|
        options[:verbose] = true
    end
end.parse!


# initialize DrQueue client
client = DrQueue.Client.new

# convert string to Python dict
pyoptions = RubyPython::PyMain.eval(options[:options], {}, {})

# initialize DrQueue job
job = DrQueue.Job.new(options[:name], options[:startframe], options[:endframe], options[:blocksize], options[:renderer], options[:scenefile], options[:retries], options[:owner], pyoptions)

# run job with client
begin
    client.job_run(job)
rescue RubyPython::PythonError => e
    puts e.message
    exit(1)
end

# tasks which have been created
tasks = client.query_task_list(job['name'])

# wait for all tasks of job to finish
if options[:wait]
    tasks.to_enum.each do |task|
        ar = client.task_wait(task['msg_id'])
        # add some verbose output
        if options[:verbose]
            cpl = ar.metadata.completed
            tmsg_id = ar.metadata.msg_id
            status = ar.status
            engine_id = ar.metadata.engine_id
            puts("Task "+tmsg_id.to_s+" finished with status '"+status.to_s+"' on engine "+engine_id.to_s+" at "+cpl.year.to_s+"-"+("%02i-%02i %02i:%02i:%02i" % [cpl.month.to_s.to_i, cpl.day.to_s.to_i, cpl.hour.to_s.to_i, cpl.minute.to_s.to_i, cpl.second.to_s.to_i])+".")
            if ar.pyerr != nil
                puts ar.pyerr
            end
        end
    end
    puts("Job %s finished." % job['name'])
end

RubyPython.stop
